File: app/src/main/java/pl/srw/billcalculator/settings/prices/PgnigPrices.java
package pl.srw.billcalculator.settings.prices;

import android.content.SharedPreferences;

import javax.inject.Inject;
import javax.inject.Singleton;

import pl.srw.billcalculator.pojo.IPgnigPrices;

@Singleton
public class PgnigPrices extends SharedPreferencesPrices implements IPgnigPrices {
    private static final String ABONAMENTOWA = "abonamentowa";
    private static final String PALIWO_GAZOWE = "paliwo_gazowe";
    private static final String DYSTRYBUCYJNA_STALA = "dystrybucyjna_stala";
    private static final String DYSTRYBUCYJNA_ZMIENNA = "dystrybucyjna_zmienna";
    private static final String WSP_KONWERSJI = "wsp_konwersji";

    private final String abonamentowa = "8.67";
    private final String paliwo_gazowe = "0.0983";
    private final String dystrybucyjna_stala = "11.39";
    private final String dystrybucyjna_zmienna = "0.02821";
    private final String wsp_konwersji = "11.290";

    @Inject
    public PgnigPrices(SharedPreferences prefs) {
        super(prefs);
    }

    public pl.srw.billcalculator.db.PgnigPrices convertToDb() {
        pl.srw.billcalculator.db.PgnigPrices dbPrices = new pl.srw.billcalculator.db.PgnigPrices();
        dbPrices.setDystrybucyjnaStala(getDystrybucyjnaStala());
        dbPrices.setDystrybucyjnaZmienna(getDystrybucyjnaZmienna());
        dbPrices.setOplataAbonamentowa(getOplataAbonamentowa());
        dbPrices.setPaliwoGazowe(getPaliwoGazowe());
        dbPrices.setWspolczynnikKonwersji(getWspolczynnikKonwersji());
        return dbPrices;
    }

    @Override
    public void clear() {
        removeOplataAbonamentowa();
        removeDystrybucyjnaStala();
        removeDystrybucyjnaZmienna();
        removePaliwoGazowe();
        removeWspolczynnikKonwersji();
    }

    @Override
    public void setDefault() {
        clear();
        setOplataAbonamentowa(getOplataAbonamentowa());
        setDystrybucyjnaStala(getDystrybucyjnaStala());
        setDystrybucyjnaZmienna(getDystrybucyjnaZmienna());
        setPaliwoGazowe(getPaliwoGazowe());
        setWspolczynnikKonwersji(getWspolczynnikKonwersji());
    }

    @Override
    public void setDefaultIfNotSet() {
        if (!containsOplataAbonamentowa())
            setDefault();
    }

    public String getOplataAbonamentowa() {
        return getPref(ABONAMENTOWA, this.abonamentowa);
    }

    public void setOplataAbonamentowa(String abonamentowa) {
        setPref(ABONAMENTOWA, abonamentowa);
    }

    public boolean containsOplataAbonamentowa() {
        return containsPref(ABONAMENTOWA);
    }

    public void removeOplataAbonamentowa() {
        removePref(ABONAMENTOWA);
    }

    public String getPaliwoGazowe() {
        return getPref(PALIWO_GAZOWE, this.paliwo_gazowe);
    }

    public void setPaliwoGazowe(String paliwoGazowe) {
        setPref(PALIWO_GAZOWE, paliwoGazowe);
    }

    public boolean containsPaliwoGazowe() {
        return containsPref(PALIWO_GAZOWE);
    }

    public void removePaliwoGazowe() {
        removePref(PALIWO_GAZOWE);
    }

    public String getDystrybucyjnaStala() {
        return getPref(DYSTRYBUCYJNA_STALA, this.dystrybucyjna_stala);
    }

    public void setDystrybucyjnaStala(String dystrybucyjnaStala) {
        setPref(DYSTRYBUCYJNA_STALA, dystrybucyjnaStala);
    }

    public boolean containsDystrybucyjnaStala() {
        return containsPref(DYSTRYBUCYJNA_STALA);
    }

    public void removeDystrybucyjnaStala() {
        removePref(DYSTRYBUCYJNA_STALA);
    }

    public String getDystrybucyjnaZmienna() {
        return getPref(DYSTRYBUCYJNA_ZMIENNA, this.dystrybucyjna_zmienna);
    }

    public void setDystrybucyjnaZmienna(String dystrybucyjnaZmienna) {
        setPref(DYSTRYBUCYJNA_ZMIENNA, dystrybucyjnaZmienna);
    }

    public boolean containsDystrybucyjnaZmienna() {
        return containsPref(DYSTRYBUCYJNA_ZMIENNA);
    }

    public void removeDystrybucyjnaZmienna() {
        removePref(DYSTRYBUCYJNA_ZMIENNA);
    }

    public String getWspolczynnikKonwersji() {
        return getPref(WSP_KONWERSJI, this.wsp_konwersji);
    }

    public void setWspolczynnikKonwersji(String wspolczynnikKonwersji) {
        setPref(WSP_KONWERSJI, wspolczynnikKonwersji);
    }

    public boolean containsWspolczynnikKonwersji() {
        return containsPref(WSP_KONWERSJI);
    }

    public void removeWspolczynnikKonwersji() {
        removePref(WSP_KONWERSJI);
    }
}

File: app/src/main/java/pl/srw/billcalculator/settings/prices/PgnigPrices.kt
package pl.srw.billcalculator.settings.prices

import android.content.SharedPreferences
import pl.srw.billcalculator.pojo.IPgnigPrices
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class PgnigPrices @Inject constructor(private val prefs: SharedPreferences)
    : SharedPrefsPrices(prefs), Restorable, IPgnigPrices {

    override var oplataAbonamentowa by stringPref(KEYS.ABONAMENTOWA)
    override var paliwoGazowe by stringPref(KEYS.PALIWO_GAZOWE)
    override var dystrybucyjnaStala by stringPref(KEYS.DYSTRYBUCYJNA_STALA)
    override var dystrybucyjnaZmienna by stringPref(KEYS.DYSTRYBUCYJNA_ZMIENNA)
    override var wspolczynnikKonwersji by stringPref(KEYS.WSP_KONWERSJI)
    override var oplataHandlowa by stringPref(KEYS.HANDLOWA)
    override var enabledOplataHandlowa by booleanPref(KEYS.HANDLOWA_ENABLED)

    fun convertToDb() = pl.srw.billcalculator.db.PgnigPrices().also {
        it.setDystrybucyjnaStala(dystrybucyjnaStala)
        it.setDystrybucyjnaZmienna(dystrybucyjnaZmienna)
        it.setOplataAbonamentowa(oplataAbonamentowa)
        it.setPaliwoGazowe(paliwoGazowe)
        it.setWspolczynnikKonwersji(wspolczynnikKonwersji)
        it.setOplataHandlowa(oplataHandlowaForDb)
    }

    override fun setDefault() {
        oplataAbonamentowa = DEFAULTS.abonamentowa
        dystrybucyjnaStala = DEFAULTS.dystrybucyjna_stala
        dystrybucyjnaZmienna = DEFAULTS.dystrybucyjna_zmienna
        paliwoGazowe = DEFAULTS.paliwo_gazowe
        wspolczynnikKonwersji = DEFAULTS.wsp_konwersji
        oplataHandlowa = DEFAULTS.handlowa
        enabledOplataHandlowa = DEFAULTS.handlowa_enabled
    }

    override fun setDefaultIfNotSet() {
        if (!prefs.contains(KEYS.ABONAMENTOWA)) {
            setDefault()
            return
        }
        if (!prefs.contains(KEYS.HANDLOWA)) {
            oplataHandlowa = DEFAULTS.handlowa
            enabledOplataHandlowa = DEFAULTS.handlowa_enabled
        }
    }

    private object DEFAULTS {
        const val abonamentowa = "8.67"
        const val paliwo_gazowe = "0.09392"
        const val dystrybucyjna_stala = "11.39"
        const val dystrybucyjna_zmienna = "0.02821"
        const val wsp_konwersji = "11.241"
        const val handlowa = "0.00"
        const val handlowa_enabled = false
    }

    private object KEYS {
        const val ABONAMENTOWA = "abonamentowa"
        const val PALIWO_GAZOWE = "paliwo_gazowe"
        const val DYSTRYBUCYJNA_STALA = "dystrybucyjna_stala"
        const val DYSTRYBUCYJNA_ZMIENNA = "dystrybucyjna_zmienna"
        const val WSP_KONWERSJI = "wsp_konwersji"
        const val HANDLOWA = "pgnig_oplata_handlowa"
        const val HANDLOWA_ENABLED = "pgnig_oplata_handlowa_enabled"
    }
}
